File: src/features/auth/auth.controller.ts
import { Controller, Post, Body, Req, Get, UseGuards, Optional, ForbiddenException } from '@nestjs/common';
import { Request } from 'express';
import { AuthService } from './auth.service';
import { LoginDto } from './dto/login.dto';
import { SessionGuard } from '../../common/guards/session.guard';
import { RegisterDto } from './dto/register.dto';
import { Roles } from '../../common/decorators/roles.decorator';
import { RolesGuard } from '../../common/guards/roles.guard';
import { VerifyEmailDto } from './dto/verify-email.dto';
import { ForgotPasswordDto } from './dto/forgot-password.dto';
import { ResetPasswordDto } from './dto/reset-password.dto';
import { Role } from '../../common/roles';
import { OrgScopeGuard } from '../../common/guards/org-scope.guard';
import { PasswordStrengthDto } from './dto/password-strength.dto';
import { LegalService } from '../legal/legal.service';
import { SessionsService } from '../sessions/sessions.service';
import { OrganizationsService } from '../organizations/organizations.service';
import { SetOrgDto } from './dto/set-org.dto';
import { UsersService } from '../users/users.service';
@Controller('auth')
export class AuthController {
  constructor(
    private auth: AuthService,
    private legal: LegalService,
    private orgs: OrganizationsService,
    private users: UsersService,
    @Optional() private sessions?: SessionsService
  ) {}

  private async regenerateSession(req: Request) {
    return new Promise<void>((resolve, reject) => {
      req.session.regenerate((err) => {
        if (err) return reject(err);
        resolve();
      });
    });
  }

  @Post('register')
  async register(@Body() dto: RegisterDto, @Req() req: Request) {
    const user = await this.auth.register(dto);
    await this.regenerateSession(req);
    const orgId = (user as any).orgId || (user as any).organizationId;
    (req as any).session.user = {
      id: (user as any)._id || (user as any).id,
      role: user.role,
      roles: (user as any).roles || [user.role],
      orgId,
      isOrgOwner: (user as any).isOrgOwner,
    };
    await this.sessions?.registerSession(req.sessionID, (user as any)._id || (user as any).id);
    const status = await this.legal.acceptanceStatus({ id: (user as any).id, orgId });
    return { user, legalRequired: status.required.length > 0, legalRequiredDocs: status.required };
  }

  @Post('login')
  async login(@Body() body: LoginDto, @Req() req: Request) {
    const user = await this.auth.login(body.email, body.password);
    await this.regenerateSession(req);
    const normalizedUser: any = {
      ...user,
      id: (user as any)._id || (user as any).id,
    };
    const orgId = normalizedUser.orgId || normalizedUser.organizationId;
    // session-based auth for dev; cookie set by express-session middleware
    (req as any).session.user = {
      id: normalizedUser.id,
      role: normalizedUser.role,
      roles: normalizedUser.roles || [normalizedUser.role],
      orgId,
      isOrgOwner: normalizedUser.isOrgOwner,
    };
    await this.sessions?.registerSession(req.sessionID, normalizedUser.id);
    const status = await this.legal.acceptanceStatus({ id: normalizedUser.id, orgId });
    return { user: normalizedUser, legalRequired: status.required.length > 0, legalRequiredDocs: status.required };
  }

  @UseGuards(SessionGuard)
  @Post('logout')
  async logout(@Req() req: Request) {
    return new Promise((resolve) => {
      const userId = req.session.user?.id;
      const sessionId = req.sessionID;
      req.session.destroy(async () => {
        if (userId && sessionId) {
          await this.sessions?.removeSession(sessionId, userId);
        }
        resolve({ status: 'ok' });
      });
    });
  }

  @UseGuards(SessionGuard)
  @Get('me')
  async me(@Req() req: Request) {
    const user = req.session.user;
    if (user?.id && !user.orgId) {
      try {
        const dbUser: any = await this.users.getById(user.id, user, true);
        const resolvedOrgId = dbUser?.orgId || dbUser?.organizationId;
        if (resolvedOrgId) {
          req.session.user = {
            ...user,
            orgId: resolvedOrgId,
            role: dbUser.role || user.role,
            roles: dbUser.roles || user.roles,
          };
        }
      } catch {
        // ignore - session remains unchanged
      }
    }
    return { user: req.session.user, legalRequired: (req as any).legalRequired || [], legalStatus: (req as any).legalStatus };
  }

  @UseGuards(SessionGuard, RolesGuard)
  @Roles(Role.SuperAdmin, Role.PlatformAdmin)
  @Post('set-org')
  async setOrg(@Body() dto: SetOrgDto, @Req() req: Request) {
    if (!req.session?.user?.id) {
      throw new ForbiddenException('Missing session');
    }
    await this.orgs.findById(dto.orgId);
    req.session.user = { ...req.session.user, orgId: dto.orgId };
    return { user: req.session.user };
  }

  @Post('verify-email')
  async verifyEmail(@Body() dto: VerifyEmailDto) {
    return this.auth.verifyEmail(dto);
  }

  @Post('password-strength')
  async passwordStrength(@Body() dto: PasswordStrengthDto) {
    return this.auth.passwordStrength(dto.password);
  }

  @Post('forgot-password')
  async forgotPassword(@Body() dto: ForgotPasswordDto) {
    return this.auth.forgotPassword(dto);
  }

  @Post('reset-password')
  async resetPassword(@Body() dto: ResetPasswordDto) {
    return this.auth.resetPassword(dto);
  }

  // Roles: admin, org owner, superadmin (tenant scoped)
  @UseGuards(SessionGuard, OrgScopeGuard, RolesGuard)
  @Roles(Role.Admin, Role.OrgAdmin, Role.SuperAdmin, Role.OrgOwner, Role.PlatformAdmin)
  @Get('users')
  async listUsers(@Req() req: Request) {
    const orgId = req.session.user?.orgId;
    return this.auth.listUsers(orgId, req.session?.user);
  }
}

File: spring-boot-api/src/main/java/com/mytegroup/api/controller/auth/AuthController.java
package com.mytegroup.api.controller.auth;

import com.mytegroup.api.dto.auth.*;
import com.mytegroup.api.entity.core.User;
import com.mytegroup.api.exception.UnauthorizedException;
import com.mytegroup.api.security.JwtTokenProvider;
import com.mytegroup.api.service.auth.AuthService;
import com.mytegroup.api.service.users.UsersService;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Authentication controller.
 * Endpoints:
 * - POST /auth/register - Public registration -> 201
 * - POST /auth/login - Public login
 * - POST /auth/logout - Authenticated logout
 * - GET /auth/me - Authenticated get current user
 * - POST /auth/verify-email - Public verify email
 * - POST /auth/password-strength - Public check password strength
 * - POST /auth/forgot-password - Public forgot password
 * - POST /auth/reset-password - Public reset password
 * - GET /auth/users - Admin+ list users
 */
@RestController
@RequestMapping("/api/auth")
@RequiredArgsConstructor
public class AuthController {

    private final AuthService authService;
    private final UsersService usersService;
    private final JwtTokenProvider jwtTokenProvider;

    @PostMapping("/register")
    @ResponseStatus(HttpStatus.CREATED)
    public Map<String, Object> register(@RequestBody @Valid RegisterDto dto) {
        User user = authService.register(
            dto.getEmail(),
            dto.getPassword(),
            dto.getUsername(),
            dto.getFirstName(),
            dto.getLastName(),
            dto.getOrganizationName(),
            dto.getOrgId(),
            Boolean.TRUE.equals(dto.getLegalAccepted()),
            dto.getInviteToken()
        );
        
        Map<String, Object> response = new HashMap<>();
        response.put("id", user.getId());
        response.put("email", user.getEmail());
        response.put("username", user.getUsername());
        response.put("isEmailVerified", user.getIsEmailVerified());
        
        return response;
    }

    @PostMapping("/login")
    public Map<String, Object> login(@RequestBody @Valid LoginDto dto) {
        User user = authService.login(dto.getEmail(), dto.getPassword());
        
        // Generate JWT token - convert User to UserDetails
        org.springframework.security.core.userdetails.UserDetails userDetails = 
            org.springframework.security.core.userdetails.User.builder()
                .username(user.getEmail())
                .password(user.getPasswordHash() != null ? user.getPasswordHash() : "")
                .authorities(user.getRoles() != null ? user.getRoles().stream()
                    .map(role -> new org.springframework.security.core.authority.SimpleGrantedAuthority("ROLE_" + role.getValue()))
                    .toArray(org.springframework.security.core.GrantedAuthority[]::new)
                    : new org.springframework.security.core.GrantedAuthority[0])
                .build();
        String token = jwtTokenProvider.generateToken(userDetails);
        
        Map<String, Object> response = new HashMap<>();
        response.put("token", token);
        response.put("user", Map.of(
            "id", user.getId(),
            "email", user.getEmail(),
            "username", user.getUsername(),
            "role", user.getRole().getValue(),
            "roles", user.getRoles().stream().map(r -> r.getValue()).toList(),
            "isEmailVerified", user.getIsEmailVerified(),
            "orgId", user.getOrganization() != null ? user.getOrganization().getId() : null
        ));
        
        return response;
    }

    @PostMapping("/logout")
    @PreAuthorize("isAuthenticated()")
    public Map<String, String> logout(HttpServletRequest request) {
        // JWT is stateless, so logout is handled client-side
        // Could optionally blacklist token here
        return Map.of("status", "ok");
    }

    @GetMapping("/me")
    @PreAuthorize("isAuthenticated()")
    public Map<String, Object> me() {
        Authentication auth = SecurityContextHolder.getContext().getAuthentication();
        if (auth == null || auth.getPrincipal() == null) {
            throw new UnauthorizedException("Not authenticated");
        }
        
        Long userId = extractUserId(auth);
        if (userId == null) {
            throw new UnauthorizedException("Invalid authentication");
        }
        
        User user = usersService.getByIdForSession(userId);
        
        Map<String, Object> response = new HashMap<>();
        response.put("id", user.getId());
        response.put("email", user.getEmail());
        response.put("username", user.getUsername());
        response.put("firstName", user.getFirstName());
        response.put("lastName", user.getLastName());
        response.put("role", user.getRole().getValue());
        response.put("roles", user.getRoles().stream().map(r -> r.getValue()).toList());
        response.put("isEmailVerified", user.getIsEmailVerified());
        response.put("isOrgOwner", user.getIsOrgOwner());
        response.put("orgId", user.getOrganization() != null ? user.getOrganization().getId() : null);
        
        return response;
    }

    @PostMapping("/verify-email")
    public Map<String, Object> verifyEmail(@RequestBody @Valid VerifyEmailDto dto) {
        User user = authService.verifyEmail(dto.getToken());
        
        Map<String, Object> response = new HashMap<>();
        response.put("id", user.getId());
        response.put("email", user.getEmail());
        response.put("isEmailVerified", user.getIsEmailVerified());
        
        return response;
    }

    @PostMapping("/password-strength")
    public Map<String, Object> passwordStrength(@RequestBody @Valid PasswordStrengthDto dto) {
        return authService.passwordStrength(dto.getPassword());
    }

    @PostMapping("/forgot-password")
    public Map<String, String> forgotPassword(@RequestBody @Valid ForgotPasswordDto dto) {
        return authService.forgotPassword(dto.getEmail());
    }

    @PostMapping("/reset-password")
    public Map<String, Object> resetPassword(@RequestBody @Valid ResetPasswordDto dto) {
        User user = authService.resetPassword(dto.getToken(), dto.getNewPassword());
        
        Map<String, Object> response = new HashMap<>();
        response.put("id", user.getId());
        response.put("email", user.getEmail());
        response.put("status", "ok");
        
        return response;
    }

    @GetMapping("/users")
    @PreAuthorize("hasAnyRole('ADMIN', 'ORG_ADMIN', 'SUPER_ADMIN', 'ORG_OWNER', 'PLATFORM_ADMIN')")
    public List<Map<String, Object>> listUsers(@RequestParam(required = false) String orgId) {
                List<User> users = authService.listUsers(orgId);
        
        return users.stream().map(user -> {
            Map<String, Object> map = new HashMap<>();
            map.put("id", user.getId());
            map.put("email", user.getEmail());
            map.put("username", user.getUsername());
            map.put("role", user.getRole().getValue());
            map.put("roles", user.getRoles().stream().map(r -> r.getValue()).toList());
            map.put("isEmailVerified", user.getIsEmailVerified());
            map.put("archivedAt", user.getArchivedAt());
            return map;
        }).toList();
    }
    
    // Helper methods
    
    private Long extractUserId(Authentication auth) {
        if (auth.getPrincipal() instanceof Long) {
            return (Long) auth.getPrincipal();
        }
        if (auth.getPrincipal() instanceof String) {
            try {
                return Long.parseLong((String) auth.getPrincipal());
            } catch (NumberFormatException e) {
                return null;
            }
        }
        return null;
    }
}
